stacks/spica/src/dashboard/pages/tutorial/tutorial.component.ts

```diff
@@ -0,0 +1,165 @@
+      );
+  }
+
+  ngOnInit() {}
+
+  ngAfterViewInit() {
+    this.stepper.steps.forEach(step => (step.completed = false));
+  }
+
+  addProperty() {
+    this.properties.push({key: "", type: "string"});
+  }
+
+  removeProperty(index: number) {
+    this.properties = this.properties.filter((_, i) => i != index);
+  }
+
+  saveSchema() {
+    if (this.isSchemaInvalid()) {
+      return;
+    }
+
+    this.bucket.properties = this.properties.reduce((acc, curr) => {
+      acc[curr.key] = {
+        title: curr.key,
+        type: this.inputTypes.indexOf(curr.type) > -1 ? curr.type : "string",
+        description: `Description of ${curr.key}`,
+        options: {
+          position: "bottom",
+          visible: true
+        }
+      };
+      return acc;
+    }, {});
+
+    this.bucket.primary = this.properties[0].key;
+
+    this.bucketService
+      .insertOne(this.bucket)
+      .toPromise()
+      .then(bucket => {
+        this.bucket = bucket;
+        this.properties.forEach(prop => (this.entry[prop.key] = undefined));
+        this.stepper.selected.completed = true;
+        this.stepper.next();
+      });
+  }
+
+  insertEntry() {
+    this.apikeyService
+      .insertOne({...emptyApiKey(), name: "tutorial"})
+      .pipe(
+        switchMap(apikey => this.apikeyService.attachPolicy("BucketFullAccess", apikey._id)),
+        switchMap(apikey =>
+          this.bucketDataService.insertOne(this.bucket._id, this.entry).pipe(
+            map(entry => {
+              return {entry: entry, apikey: apikey};
+            })
+          )
+        )
+      )
+      .toPromise()
+      .then(({entry, apikey}) => {
+        this.entry = entry;
+
+        this.curl = this.createCURL(this.bucket._id, entry._id, apikey.key);
+        this.exampleInfo.url = this.createUrl(this.bucket._id, entry._id);
+        this.exampleInfo.headers.Authorization = `APIKEY ${apikey.key}`;
+
+        this.stepper.selected.completed = true;
+        this.stepper.next();
+        this._snackBar.open(
+          "Congratulations. You have been completed the tutorial. Now, you can continue to develop your project.",
+          "",
+          {
+            duration: 5000
+          }
+        );
+      });
+  }
+
+  createUrl(bucketId: string, entryId: string) {
+    return `${environment.api}/bucket/${bucketId}/data/${entryId}`;
+  }
+
+  createCURL(bucketId: string, entryId: string, apikey: string) {
+    if (!environment.api || !bucketId || !entryId || !apikey) {
+      return;
+    }
+    let url = this.createUrl(bucketId, entryId);
+    return `curl -H "Authorization:APIKEY ${apikey}" ${url}`;
+  }
+
+  isSchemaInvalid() {
+    let hasUndefinedKey = false;
+    let hasDuplicatedKey = false;
+    this.properties.reduce((acc, curr) => {
+      if (!curr.key) {
+        hasUndefinedKey = true;
+      } else if (acc.includes(curr.key)) {
+        hasDuplicatedKey = true;
+      } else {
+        acc.push(curr.key);
+      }
+      return acc;
+    }, []);
```

Comment:
How about we just use a simple plain "for of" loop here? I think using "reduce" here has no benefit.